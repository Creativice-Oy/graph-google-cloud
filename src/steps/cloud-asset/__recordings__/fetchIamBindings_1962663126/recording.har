{
  "log": {
    "_recordingName": "fetchIamBindings",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 739,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "739"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 300,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 581,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 581,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Mon, 07 Jun 2021 22:21:18 GMT"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 511,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-07T22:21:18.300Z",
        "time": 190,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 190
        }
      },
      {
        "_id": "90071867bd49fb876fbfa88d098851a2",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/5.0.2 gl-node/14.15.0 auth/7.1.0"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/5.0.2 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "cloudasset.googleapis.com"
            }
          ],
          "headersSize": 571,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "pageSize",
              "value": "500"
            }
          ],
          "url": "https://cloudasset.googleapis.com/v1/organizations/958457776463:searchAllIamPolicies?pageSize=500"
        },
        "response": {
          "bodySize": 3094,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 3094,
            "text": "{\"results\":[{\"resource\":\"//cloudresourcemanager.googleapis.com/folders/430707931057\",\"policy\":{\"bindings\":[{\"role\":\"roles/resourcemanager.folderAdmin\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]},{\"role\":\"roles/resourcemanager.folderEditor\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/folders/811983343328\",\"policy\":{\"bindings\":[{\"role\":\"roles/resourcemanager.folderAdmin\",\"members\":[\"user:viragsf@gmail.com\"]},{\"role\":\"roles/resourcemanager.folderEditor\",\"members\":[\"user:viragsf@gmail.com\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/folders/1065006483234\",\"policy\":{\"bindings\":[{\"role\":\"roles/resourcemanager.folderAdmin\",\"members\":[\"user:viragsf@gmail.com\"]},{\"role\":\"roles/resourcemanager.folderEditor\",\"members\":[\"user:viragsf@gmail.com\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/folders/299817291758\",\"policy\":{\"bindings\":[{\"role\":\"roles/resourcemanager.folderAdmin\",\"members\":[\"user:viragsf@gmail.com\"]},{\"role\":\"roles/resourcemanager.folderEditor\",\"members\":[\"user:viragsf@gmail.com\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/projects/intense-pointer-315514\",\"project\":\"projects/592147458023\",\"policy\":{\"bindings\":[{\"role\":\"roles/owner\",\"members\":[\"user:nick.dowmon@jupiterone.dev\"]}]}},{\"resource\":\"//bigquery.googleapis.com/projects/j1-gc-integration-dev-v3/datasets/natality\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/bigquery.dataEditor\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/bigquery.dataOwner\",\"members\":[\"projectOwner:j1-gc-integration-dev-v3\",\"user:viragsf@gmail.com\"]},{\"role\":\"roles/bigquery.dataViewer\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]}]}},{\"resource\":\"//bigquery.googleapis.com/projects/j1-gc-integration-dev-v3/datasets/cmed_example_dataset\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/bigquery.dataEditor\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/bigquery.dataOwner\",\"members\":[\"projectOwner:j1-gc-integration-dev-v3\",\"serviceAccount:j1-gc-integration-dev-sa-tf@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/bigquery.dataViewer\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]}]}},{\"resource\":\"//storage.googleapis.com/us.artifacts.j1-gc-integration-dev-v3.appspot.com\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/storage.legacyBucketOwner\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\",\"projectOwner:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/storage.legacyBucketReader\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]}]}},{\"resource\":\"//storage.googleapis.com/gcf-sources-167984947943-us-central1\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/storage.legacyBucketOwner\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\",\"projectOwner:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/storage.legacyBucketReader\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/storage.legacyObjectOwner\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\",\"projectOwner:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/storage.legacyObjectReader\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]}]}},{\"resource\":\"//cloudkms.googleapis.com/projects/j1-gc-integration-dev-v3/locations/us/keyRings/j1-gc-integration-dev-v3-disk-bucket-ring/cryptoKeys/j1-gc-integration-dev-v3-disk-bucket-key\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/cloudkms.cryptoKeyEncrypterDecrypter\",\"members\":[\"serviceAccount:service-167984947943@compute-system.iam.gserviceaccount.com\"]}]}},{\"resource\":\"//storage.googleapis.com/j1-gc-integration-dev-v3-sink-logging-bucket\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/storage.legacyBucketOwner\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\",\"projectOwner:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/storage.legacyBucketReader\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]}]}},{\"resource\":\"//bigquery.googleapis.com/projects/j1-gc-integration-dev-v3/datasets/test_big_query_dataset\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/bigquery.dataOwner\",\"members\":[\"serviceAccount:j1-gc-integration-dev-sa-tf@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]}]}},{\"resource\":\"//storage.googleapis.com/j1-gc-integration-dev-v3cloudfunctions\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/storage.legacyBucketOwner\",\"members\":[\"projectEditor:j1-gc-integration-dev-v3\",\"projectOwner:j1-gc-integration-dev-v3\"]},{\"role\":\"roles/storage.legacyBucketReader\",\"members\":[\"projectViewer:j1-gc-integration-dev-v3\"]}]}},{\"resource\":\"//cloudkms.googleapis.com/projects/j1-gc-integration-dev-v3/locations/us-central1/keyRings/j1-gc-integration-dev-v3-encrypted-sql-instance-ring/cryptoKeys/j1-gc-integration-dev-v3-encrypted-sql-instance-key\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/cloudkms.cryptoKeyEncrypterDecrypter\",\"members\":[\"serviceAccount:service-167984947943@gcp-sa-cloud-sql.iam.gserviceaccount.com\"]}]}},{\"resource\":\"//cloudkms.googleapis.com/projects/j1-gc-integration-dev-v3/locations/us/keyRings/example-keyring\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"roles/cloudkms.cryptoKeyEncrypterDecrypter\",\"members\":[\"serviceAccount:bq-167984947943@bigquery-encryption.iam.gserviceaccount.com\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/projects/learned-hour-315416\",\"project\":\"projects/988862309073\",\"policy\":{\"bindings\":[{\"role\":\"roles/owner\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]}]}},{\"resource\":\"//cloudbilling.googleapis.com/billingAccounts/01694E-12D60B-BEBCE5\",\"policy\":{\"bindings\":[{\"role\":\"roles/billing.admin\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/projects/winged-guild-315415\",\"project\":\"projects/619127027446\",\"policy\":{\"bindings\":[{\"role\":\"roles/owner\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/projects/j1-gc-integration-dev-v3\",\"project\":\"projects/167984947943\",\"policy\":{\"bindings\":[{\"role\":\"projects/j1-gc-integration-dev-v3/roles/167984947943customrole\",\"members\":[\"serviceAccount:j1-gc-integration-dev-sa-tf@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"projects/j1-gc-integration-dev-v3/roles/167984947943customroleconditions\",\"members\":[\"serviceAccount:j1-gc-integration-dev-sa-tf@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"],\"condition\":{\"expression\":\"resource.name != \\\"bogusunknownresourcename\\\"\",\"title\":\"Test condition title\",\"description\":\"Test condition description\"}},{\"role\":\"projects/j1-gc-integration-dev-v3/roles/167984947943extendedrole\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/bigquery.dataViewer\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/bigquery.resourceViewer\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/binaryauthorization.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@gcp-sa-binaryauthorization.iam.gserviceaccount.com\"]},{\"role\":\"roles/cloudbuild.builds.builder\",\"members\":[\"serviceAccount:167984947943@cloudbuild.gserviceaccount.com\"]},{\"role\":\"roles/cloudbuild.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@gcp-sa-cloudbuild.iam.gserviceaccount.com\"]},{\"role\":\"roles/cloudfunctions.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@gcf-admin-robot.iam.gserviceaccount.com\"]},{\"role\":\"roles/cloudkms.cryptoKeyDecrypter\",\"members\":[\"serviceAccount:service-167984947943@gcp-sa-pubsub.iam.gserviceaccount.com\"]},{\"role\":\"roles/cloudkms.cryptoKeyEncrypter\",\"members\":[\"serviceAccount:service-167984947943@gcp-sa-pubsub.iam.gserviceaccount.com\"]},{\"role\":\"roles/compute.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@compute-system.iam.gserviceaccount.com\"]},{\"role\":\"roles/container.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@container-engine-robot.iam.gserviceaccount.com\"]},{\"role\":\"roles/containeranalysis.ServiceAgent\",\"members\":[\"serviceAccount:service-167984947943@container-analysis.iam.gserviceaccount.com\"]},{\"role\":\"roles/containerregistry.ServiceAgent\",\"members\":[\"serviceAccount:service-167984947943@containerregistry.iam.gserviceaccount.com\"]},{\"role\":\"roles/editor\",\"members\":[\"serviceAccount:167984947943-compute@developer.gserviceaccount.com\",\"serviceAccount:167984947943@cloudservices.gserviceaccount.com\",\"serviceAccount:j1-gc-integration-dev-v3@appspot.gserviceaccount.com\"]},{\"role\":\"roles/iam.securityReviewer\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/memcache.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@cloud-memcache-sa.iam.gserviceaccount.com\"]},{\"role\":\"roles/owner\",\"members\":[\"serviceAccount:j1-gc-integration-dev-sa-tf@j1-gc-integration-dev-v3.iam.gserviceaccount.com\",\"user:austin.kelleher@jupiterone.dev\",\"user:michael.knoedel@jupiterone.com\"]},{\"role\":\"roles/pubsub.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@gcp-sa-pubsub.iam.gserviceaccount.com\"]},{\"role\":\"roles/redis.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@cloud-redis.iam.gserviceaccount.com\"]},{\"role\":\"roles/run.serviceAgent\",\"members\":[\"serviceAccount:service-167984947943@serverless-robot-prod.iam.gserviceaccount.com\"]},{\"role\":\"roles/viewer\",\"members\":[\"group:integrations@jupiterone.dev\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/folders/306329820365\",\"policy\":{\"bindings\":[{\"role\":\"roles/resourcemanager.folderAdmin\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]},{\"role\":\"roles/resourcemanager.folderEditor\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]}]}},{\"resource\":\"//cloudresourcemanager.googleapis.com/organizations/958457776463\",\"policy\":{\"bindings\":[{\"role\":\"roles/billing.creator\",\"members\":[\"domain:jupiterone.dev\"]},{\"role\":\"roles/cloudasset.viewer\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/iam.securityReviewer\",\"members\":[\"serviceAccount:j1-gc-integration-dev-org@my-j1-test-proj-test.iam.gserviceaccount.com\",\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/owner\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\",\"user:adam.williams@jupiterone.dev\",\"user:austin.kelleher@jupiterone.dev\",\"user:michael.knoedel@jupiterone.com\",\"user:nick.dowmon@jupiterone.dev\",\"user:viragsf@gmail.com\"]},{\"role\":\"roles/resourcemanager.folderAdmin\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]},{\"role\":\"roles/resourcemanager.folderCreator\",\"members\":[\"user:viragsf@gmail.com\"]},{\"role\":\"roles/resourcemanager.organizationAdmin\",\"members\":[\"user:adam.williams@jupiterone.dev\",\"user:austin.kelleher@jupiterone.dev\",\"user:michael.knoedel@jupiterone.com\",\"user:nick.dowmon@jupiterone.dev\",\"user:viragsf@gmail.com\"]},{\"role\":\"roles/resourcemanager.organizationViewer\",\"members\":[\"serviceAccount:j1-gc-integration-dev-v3@j1-gc-integration-dev-v3.iam.gserviceaccount.com\"]},{\"role\":\"roles/resourcemanager.projectCreator\",\"members\":[\"domain:jupiterone.dev\",\"user:austin.kelleher@jupiterone.dev\"]},{\"role\":\"roles/resourcemanager.projectMover\",\"members\":[\"user:austin.kelleher@jupiterone.dev\"]}]}}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Mon, 07 Jun 2021 22:21:18 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 488,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-07T22:21:18.502Z",
        "time": 335,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 335
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
