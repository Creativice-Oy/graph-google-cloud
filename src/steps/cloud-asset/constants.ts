import {
  generateRelationshipType,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { ANY_RESOURCE } from '../../constants';
import {
  GOOGLE_GROUP_ENTITY_TYPE,
  GOOGLE_USER_ENTITY_TYPE,
  IAM_ROLE_ENTITY_TYPE,
  IAM_SERVICE_ACCOUNT_ENTITY_TYPE,
} from '../iam';

export const CLOUD_ASSET_STEPS = {
  BINDINGS: 'fetch-iam-bindings',
};

export const bindingEntities = {
  BINDINGS: {
    _type: 'google_iam_binding',
    _class: ['AccessPolicy'], // Technically not a policy, but functions just like one. IAM Policies are groups of IAM bindings.
    resourceName: 'IAM Binding',
  },
};

const IAM_PRINCIPAL_TYPES = [
  // 'google_domain', // These are getting ignored for now
  IAM_SERVICE_ACCOUNT_ENTITY_TYPE,
  GOOGLE_GROUP_ENTITY_TYPE,
  GOOGLE_USER_ENTITY_TYPE,
];

export const BINDING_ASSIGNED_PRINCIPAL_RELATIONSHIPS = IAM_PRINCIPAL_TYPES.map(
  (principalType) => {
    return {
      _type: generateRelationshipType(
        RelationshipClass.ASSIGNED,
        bindingEntities.BINDINGS._type,
        principalType,
      ),
      sourceType: bindingEntities.BINDINGS._type,
      _class: RelationshipClass.ASSIGNED,
      targetType: principalType,
    };
  },
);

export const PRINCIPAL_ASSIGNED_ROLE_RELATIONSHIPS = IAM_PRINCIPAL_TYPES.map(
  (principalType) => {
    return {
      _type: generateRelationshipType(
        RelationshipClass.ASSIGNED,
        principalType,
        IAM_ROLE_ENTITY_TYPE,
      ),
      sourceType: principalType,
      _class: RelationshipClass.ASSIGNED,
      targetType: IAM_ROLE_ENTITY_TYPE,
    };
  },
);

/**
 * IAM policies can target any resource in Google Cloud. Because we do not ingest every resource,
 * we have chosen, instead, to represent the relationship as IAM Binding assigned to ANY_RESOURCE.
 */
export const BINDING_ASSIGNED_ANY_RESOURCE_RELATIONSHIP = {
  _type: generateRelationshipType(
    RelationshipClass.ASSIGNED,
    bindingEntities.BINDINGS._type,
    ANY_RESOURCE,
  ),
  sourceType: bindingEntities.BINDINGS._type,
  _class: RelationshipClass.ASSIGNED,
  targetType: ANY_RESOURCE,
};
