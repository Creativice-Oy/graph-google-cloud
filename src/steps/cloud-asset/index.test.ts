import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../..';
import { integrationConfig } from '../../../test/config';
import { withRecording } from '../../../test/recording';
import { fetchIamBindings } from '.';
import { bindingEntities } from './constants';
import {
  buildOrgFolderProjectMappedRelationships,
  fetchResourceManagerFolders,
  fetchResourceManagerOrganization,
} from '../resource-manager';
import {
  fetchIamCustomRoles,
  fetchIamManagedRoles,
  fetchIamServiceAccounts,
} from '../iam';
import { Entity, Relationship } from '@jupiterone/integration-sdk-core';
import { filterGraphObjects } from '../../../test/helpers/filterGraphObjects';

describe('#fetchIamBindings', () => {
  function separateGraphObjectsByType<T extends Entity | Relationship>(
    collected: T[],
    encounteredTypes: string[],
  ) {
    const relationshipsByType: Record<string, T[]> = {};
    let rest: T[] = collected;
    for (const type of encounteredTypes) {
      const filterResult = filterGraphObjects(rest, (o) => o._type === type);
      rest = filterResult.rest;
      relationshipsByType[type] = filterResult.targets;
    }
    return relationshipsByType;
  }

  test('should collect data', async () => {
    await withRecording('fetchIamBindings', __dirname, async () => {
      const context = createMockStepExecutionContext<IntegrationConfig>({
        // Temporary tweak to make this test pass since its recording has been updated from the new organization/v3
        instanceConfig: {
          ...integrationConfig,
          serviceAccountKeyFile:
            integrationConfig.serviceAccountKeyFile.replace(
              'j1-gc-integration-dev-v2',
              'j1-gc-integration-dev-v3',
            ),
          serviceAccountKeyConfig: {
            ...integrationConfig.serviceAccountKeyConfig,
            project_id: 'j1-gc-integration-dev-v3',
          },
        },
      });

      await fetchResourceManagerOrganization(context);
      await fetchResourceManagerFolders(context);
      await buildOrgFolderProjectMappedRelationships(context);
      await fetchIamCustomRoles(context);
      await fetchIamManagedRoles(context);
      await fetchIamServiceAccounts(context);
      await fetchIamBindings(context);

      expect({
        numCollectedEntities: context.jobState.collectedEntities.length,
        numCollectedRelationships:
          context.jobState.collectedRelationships.length,
        collectedEntities: context.jobState.collectedEntities.length,
        collectedRelationships: context.jobState.collectedRelationships.length,
        encounteredTypes: context.jobState.encounteredTypes,
      }).toMatchSnapshot();

      // Relationships
      const {
        google_iam_binding_uses_role,
        google_iam_binding_assigned_user,
        google_iam_binding_assigned_group,
        google_iam_binding_assigned_service_account,
        google_user_assigned_iam_role,
        google_group_assigned_iam_role,
        google_iam_service_account_assigned_role,
      } = separateGraphObjectsByType(
        context.jobState.collectedRelationships,
        context.jobState.encounteredTypes,
      );

      // Each of these relationships could either be mapped or direct depending on if it has been previously ingested in this run so not checking if direct or mapped here
      expect(google_iam_binding_uses_role.length).toBeGreaterThan(0);
      expect(google_iam_binding_assigned_user.length).toBeGreaterThan(0);
      expect(google_iam_binding_assigned_group.length).toBeGreaterThan(0);
      expect(
        google_iam_binding_assigned_service_account.length,
      ).toBeGreaterThan(0);
      expect(google_user_assigned_iam_role.length).toBeGreaterThan(0);
      expect(google_group_assigned_iam_role.length).toBeGreaterThan(0);
      expect(google_iam_service_account_assigned_role.length).toBeGreaterThan(
        0,
      );

      // Entities
      const { google_iam_binding, google_iam_role } =
        separateGraphObjectsByType(
          context.jobState.collectedEntities,
          context.jobState.encounteredTypes,
        );

      expect(google_iam_binding.length).toBeGreaterThan(0);
      expect(google_iam_binding).toMatchGraphObjectSchema({
        _class: bindingEntities.BINDINGS._class,
        schema: {
          properties: {
            _type: { const: bindingEntities.BINDINGS._type },
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            resource: { type: 'string' },
            projectId: { type: 'string' },
            members: { type: 'array' },
            'condition.title': { type: 'string' },
            'condition.description': { type: 'string' },
            'condition.expression': { type: 'string' },
          },
        },
      });

      expect(google_iam_role.length).toBeGreaterThan(0);
      expect(google_iam_role).toMatchGraphObjectSchema({
        _class: ['AccessRole'],
        schema: {
          additionalProperties: false,
          properties: {
            _type: { const: 'google_iam_role' },
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            description: { type: 'string' },
            stage: { type: 'string' },
            custom: { type: 'boolean' },
            deleted: { type: 'boolean' },
            permissions: {
              type: 'array',
              items: { type: 'string' },
            },
            etag: { type: 'string' },
            readonly: { type: 'boolean' },
          },
        },
      });
    });
  });
});
