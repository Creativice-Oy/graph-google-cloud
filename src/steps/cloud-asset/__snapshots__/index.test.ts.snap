// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchIamBindings should collect data 1`] = `
Object {
  "collectedEntities": 144,
  "collectedRelationships": 307,
  "encounteredTypes": Array [
    "google_cloud_organization",
    "google_cloud_folder",
    "google_cloud_organization_has_folder",
    "google_cloud_folder_has_folder",
    "google_organization_has_project",
    "google_folder_has_project",
    "google_iam_role",
    "google_iam_service_account",
    "google_iam_service_account_key",
    "google_iam_service_account_has_key",
    "google_iam_binding",
    "google_iam_binding_assigned_user",
    "google_user_assigned_iam_role",
    "google_iam_binding_assigned_service_account",
    "google_iam_service_account_assigned_role",
    "google_iam_binding_assigned_group",
    "google_group_assigned_iam_role",
    "google_iam_binding_uses_role",
    "google_iam_binding_allows_ANY_RESOURCE",
    "google_iam_binding_allows_cloud_folder",
    "google_iam_binding_allows_cloud_organization",
  ],
  "numCollectedEntities": 144,
  "numCollectedRelationships": 307,
}
`;
